File: src/Scripting/Lua/flyweight.lua
local StyleFactory = {}
StyleFactory.__index = StyleFactory

function StyleFactory.new()
  return setmetatable({ styles = {}, count = 0 }, StyleFactory)
end

function StyleFactory:get(font, size, color)
  local key = font .. "|" .. size .. "|" .. color
  local style = self.styles[key]
  if not style then
    style = { font = font, size = size, color = color }
    self.styles[key] = style
    self.count = self.count + 1
  end
  return style
end

local factory = StyleFactory.new()
local red1 = factory:get("Inter", 12, "red")
local red2 = factory:get("Inter", 12, "red")
factory:get("Inter", 12, "blue")

print(string.format("styles=%d;shared=%s;text=ABC", factory.count, tostring(red1 == red2)))

File: src/Scripting/Perl/flyweight.pl
use strict;
use warnings;
use Scalar::Util qw(refaddr);

my %styles;
sub get_style {
    my ($font, $size, $color) = @_;
    my $key = join '|', $font, $size, $color;
    $styles{$key} //= { font => $font, size => $size, color => $color };
    return $styles{$key};
}

my $red1 = get_style('Inter', 12, 'red');
my $red2 = get_style('Inter', 12, 'red');
get_style('Inter', 12, 'blue');
my $shared = refaddr($red1) == refaddr($red2) ? 'true' : 'false';
print 'styles=' . scalar(keys %styles) . ";shared=$shared;text=ABC\n";
